File: src/check.c
/*
 * check.c
 *
 * Copyright (C) 2003 Bastian Blank <waldi@debian.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <config.h>

#include <sys/stat.h>
#include <sys/types.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "check.h"
#include "execute.h"
#include "frontend.h"
#include "suite.h"

/* Length of a SHA256 hash in hex representation */
#define SHA256_HEX_LENGTH 64

static int check_sum (const char *target, const char *exec, const char *sum, const char *message)
{
  int ret;
  FILE *in;
  char buf[1024];

  log_message (LOG_MESSAGE_INFO_DOWNLOAD_VALIDATE, message);

  snprintf (buf, sizeof (buf), "%s %s", exec, target);

  log_text (DI_LOG_LEVEL_DEBUG, "Execute \"%s\"", buf);
  in = popen (buf, "r");
  if (!fgets (buf, sizeof (buf), in))
    return 1;

  ret = pclose (in);
  if (ret)
    return 1;

  if (!strncmp (buf, sum, SHA256_HEX_LENGTH))
    return 0;
  return 1;
}

int check_deb (const char *target, di_package *p, const char *message)
{
  return check_sum (target, "sha256sum", p->sha256, message);
}

int check_packages (const char *target, const char *ext, di_release *rel)
{
  char buf_name[64];
  char buf_file[128];
  di_release_file *item;
  di_rstring key;

  snprintf (buf_name, sizeof (buf_name), "Packages%s", ext);
  snprintf (buf_file, sizeof (buf_file), "main/binary-%s/Packages%s", arch, ext);
  key.string = (char *) buf_file;
  key.size = strlen (buf_file);
  item = di_hash_table_lookup (rel->sha256, &key);
  if (!item)
    log_text (DI_LOG_LEVEL_ERROR, "Can't find checksum for Packages file");

  if (item->sum[1])
    return check_sum (target, "sha256sum", item->sum[1], buf_name);
  return 1;
}

File: src/check.zig
const std = @import("std");
const log = std.log.scoped(.check);
const mem = std.mem;
const Sha256 = std.crypto.hash.sha2.Sha256;
const assert = std.debug.assert;
const c = @import("c");
const buf_size = 4 * 1024;

fn checkSum(
    target_maybe: ?[*:0]const u8,
    sum_maybe: ?[*:0]const u8,
    message: ?[*:0]const u8,
) bool {
    return checkSumAllowFail(target_maybe, sum_maybe, message) catch |err| {
        log.err("unexpected checksum failure for {s}: {t}", .{ target_maybe.?, err });
        return false;
    };
}

fn checkSumAllowFail(
    target_maybe: ?[*:0]const u8,
    sum_maybe: ?[*:0]const u8,
    message: ?[*:0]const u8,
) !bool {
    assert(sum_maybe != null);
    assert(target_maybe != null);
    const sum = mem.span(sum_maybe.?);
    const target = mem.span(target_maybe.?);
    c.log_message(c.LOG_MESSAGE_INFO_DOWNLOAD_VALIDATE, message);

    var file = try std.fs.cwd().openFile(target, .{});
    defer file.close();

    var buf_reader: [buf_size]u8 = undefined;
    var buf_hasher: [buf_size]u8 = undefined;

    var reader = file.reader(&buf_reader);
    var sha256_writer: std.Io.Writer.Hashing(Sha256) = .init(&buf_hasher);
    _ = try reader.interface.streamRemaining(&sha256_writer.writer);
    sha256_writer.writer.flush() catch unreachable;
    const target_hash = sha256_writer.hasher.finalResult();
    const hash_hex = std.fmt.bytesToHex(target_hash, .lower);
    return mem.eql(u8, &hash_hex, sum[0..hash_hex.len]);
}

export fn check_deb(
    target: ?[*:0]const u8,
    package: ?*c.di_package,
    message: ?[*:0]const u8,
) c_int {
    assert(package != null);
    return @intFromBool(!checkSum(target, package.?.sha256, message));
}

export fn check_packages(
    target: ?[*:0]const u8,
    ext: ?[*:0]const u8,
    rel: ?*c.di_release,
) c_int {
    assert(target != null);
    assert(ext != null);
    assert(rel != null);

    var buf_name: [64]u8 = undefined;
    var buf_file: [128]u8 = undefined;
    const name = std.fmt.bufPrintZ(&buf_name, "Packages{s}", .{ext.?}) catch |err| {
        log.err("Can't format name: {t}", .{err});
        return 1;
    };
    const file = std.fmt.bufPrintZ(&buf_file, "main/binary-{s}/Packages{s}", .{ c.arch, ext.? }) catch |err| {
        log.err("Can't format file: {t}", .{err});
        return 1;
    };

    const key: c.di_rstring = .{
        .size = @intCast(file.len),
        .string = file.ptr,
    };

    const item: ?*c.di_release_file = @ptrCast(@alignCast(c.di_hash_table_lookup(rel.?.sha256, &key)));
    if (item == null) {
        c.log_text(c.DI_LOG_LEVEL_ERROR, "Can't find checksum for Packages file");
        return 1;
    }

    if (item.?.sum[1] != null) {
        return @intFromBool(!checkSum(target, item.?.sum[1], name.ptr));
    }

    return 1;
}
